File: python/23_non_abundant_sums.py
from math import sqrt

def non_abundant_sums() -> int:

    sum: int = 0
    nums = set() # faster for checking membership than a list
    # all integers greater than 28123 can be written as the sum of 2 abundant numbers
    for i in range(1, 28124):
        # add to abundant number set
        if divisor_sum(i) > i:
            nums.add(i)
        if not is_abundant_sum(nums, i):
            sum += i
        
    return sum
    
def is_abundant_sum(nums: set, n: int) -> bool:
    # check if difference of n and any abundant number in the set
    # is also an abundant number, then n is an abundant sum
    for num in nums:
        if (n - num) in nums:
            return True
    return False
    
def divisor_sum(num: int) -> int:
    sum: int = 1
    for i in range(2, int(sqrt(num)) + 1):
        if num % i == 0:
            sum += i + num // i
    
    if int(sqrt(num)) == sqrt(num):
        sum -= int(sqrt(num))
    return sum

if __name__ == "__main__":
    print(non_abundant_sums())

File: java/_23_NonAbundantSums.java
import java.util.*;

public class _23_NonAbundantSums{
    
    // Keep track of discovered abundant numbers
    static HashSet<Integer> abNums = new HashSet<>();
        
    public static int sumProperDivs(int num){
        // Proper divisors are strictly less than the dividend.
        int factorSum = 1; 
        for (int i = 2; i <= Math.sqrt(num); i++){
            if (num % i == 0){
                factorSum += i + num / i;
            }
        }
        if (Math.floor(Math.sqrt(num)) * Math.floor(Math.sqrt(num)) == num){
            factorSum -= Math.sqrt(num);
        }
        return factorSum;
    }
    
    public static boolean isAbSum(int num){
        // Instead of checking every possible sum of two abundant numbers (0(n^2)),
        // check if the current number minus any of the previously found
        // abundant numbers is also an abundant number (0(n)).
        for (int abNum : abNums){
            if (abNums.contains(num - abNum)){
                return true;
            }
        }
        return false;
    }
    
    public static void main(String[] args){
        int sum = 0;
        
        for (int i = 1; i < 28123; i++){
            int divSum = sumProperDivs(i);
            
            if (divSum > i){
                abNums.add(i);
            }
            if (!isAbSum(i)){
                sum += i;
            }
        }
        
        System.out.println("The sum of all positive integers under 28123 that cannot be written as the sum of two abundant numbers is " + sum); // 4179871
    }
}